### PR TITLE
[ebpf] Add support for RTT (round-trip time)

### PR DESCRIPTION
**Pull request by p-lambert**
### What does this PR do?

Depends on https://github.com/DataDog/datadog-agent/pull/3969

- [x] Enables offset guessing for RTT fields.
- [x] Add RTT to `ConnectionStats`
- [x] Display RTT on `nettop`

OBS.: This PR does not update the connection model used in the agent. This will be done in another PR.

### CODE REVIEW

File: pkg/ebpf/offsetguess.go [hunk 1]
```diff
@@ -218,6 +267,26 @@ func checkAndUpdateCurrentOffset(module *elf.Module, mp *elf.Map, status *tracer
 			status.offset_ino = 0
 			status.offset_netns++
 		}
+	case guessRTT:
+		if status.rtt == C.__u32(expected.rtt) && status.rtt_var == C.__u32(expected.rttVar) {
+			status.what = guessDaddrIPv6
+			break
+		}
+		// We know that these two fields are always next to each other, 4 bytes apart:
+		// https://elixir.bootlin.com/linux/v4.6/source/include/linux/tcp.h#L232
```
Comment (sunhay) [inline pkg/ebpf/offsetguess.go]: Have you confirmed this on all versions between 4.4 -> 5.2?

Comment (sunhay) [reply pkg/ebpf/offsetguess.go]: (specifically just incrementing the minor version and confirming would be good)

File: pkg/ebpf/offsetguess.go [hunk 2]
```diff
@@ -218,6 +267,26 @@ func checkAndUpdateCurrentOffset(module *elf.Module, mp *elf.Map, status *tracer
...
+		// rtt -> srtt_us
+		// rtt_var -> mdev_us
+		status.offset_rtt++
+		status.offset_rtt_var = status.offset_rtt + 4
+
+		// For now we'll tolerate the case where we can't find the offsets for RTT metrics
+		if status.offset_rtt > thresholdInetSock {
+			log.Warn("could not guess offsets for RTT fields. moving on.")
```
Comment (sunhay) [inline pkg/ebpf/offsetguess.go]: `TCP RTT`*

File: pkg/ebpf/c/tracer-ebpf.c [hunk 3]
```diff
@@ -290,6 +296,14 @@ static int guess_offsets(tracer_status_t* status, struct sock* skp) {
         }
         new_status.netns = possible_netns;
         break;
+    case GUESS_RTT:
+        possible_rtt = 0;
+        possible_rtt_var = 0;
+        bpf_probe_read(&possible_rtt, sizeof(possible_rtt), ((char*)skp) + status->offset_rtt);
+        bpf_probe_read(&possible_rtt_var, sizeof(possible_rtt_var), ((char*)skp) + status->offset_rtt_var);
+        new_status.rtt = possible_rtt;
```
Comment (sunhay) [inline pkg/ebpf/c/tracer-ebpf.c]: Any reason not to read this value directly into `new_status`?

Comment (p-lambert) [reply pkg/ebpf/c/tracer-ebpf.c]: no, I was just following the same pattern present in all other cases from this big `switch` statement, but now that you brought this up I think I'll add another commit to my refactoring PR and clean up this function. there is a a good chunk of useless boilerplate code there that we can get rid of.

File: pkg/ebpf/offsetguess.go (hunk 1 above)
Comment (p-lambert) [reply pkg/ebpf/offsetguess.go]: Yes, that's the case for >= v3.0

File: pkg/ebpf/offsetguess.go [hunk 4]
```diff
@@ -37,6 +37,11 @@ const (
 	// The source port is much further away in the inet sock.
 	thresholdInetSock = 2000
 
+	// When GETSOCKOPT is called the TCP socket field representing rttVar (mdev_us)
+	// loses 2 bits of precision due to the operation (mdev_us >> 2) << 2, so we use
+	// this mask to correct that imprecision when comparing the values.
+	rttVarMask = ^C.__u32(3)
```
Comment (sunhay) [inline pkg/ebpf/offsetguess.go]: You probably already did, but we should make sure we guarantee that this holds true for most/all kernel versions we care about :)

File: pkg/ebpf/event_common.go [hunk 5]
```diff
@@ -108,8 +112,17 @@ func (c ConnectionStats) String() string {
 		c.Direction,
 		c.MonotonicSentBytes, c.LastSentBytes,
 		c.MonotonicRecvBytes, c.LastRecvBytes,
-		c.MonotonicRetransmits, c.LastRetransmits,
 	)
+
+	if c.Type == TCP {
+		str += fmt.Sprintf(
```
Comment (p-lambert) [inline pkg/ebpf/event_common.go]: A string builder wouldn't help here, so I'm just concatenating the strings.

File: pkg/ebpf/c/tracer-ebpf.c [hunk 6]
```diff
@@ -484,16 +430,33 @@ static int handle_message(struct sock* sk,
...
+    update_tcp_stats(&t, stats);
 
     // Update latest timestamp that we've seen - for connection expiration tracking
-    u64 zero = 0;
     bpf_map_update_elem(&latest_ts, &zero, &ts, BPF_ANY);
     return 0;
 }
 
+__attribute__((always_inline))
+static void handle_tcp_stats(conn_tuple_t* t, tracer_status_t* status, struct sock* sk) {
```
Comment (p-lambert) [inline pkg/ebpf/c/tracer-ebpf.c]: In a future PR we could track the retransmits here (by also guessing the retransmit field in the TCP socket struct) and remove the associated code/kprobe.

File: pkg/ebpf/c/tracer-ebpf.c [hunk 7]
```diff
@@ -547,7 +510,13 @@ int kprobe__tcp_sendmsg(struct pt_regs* ctx) {
     }
     log_debug("kprobe/tcp_sendmsg: pid_tgid: %d, size: %d\n", pid_tgid, size);
 
-    return handle_message(sk, status, pid_tgid, CONN_TYPE_TCP, size, 0);
+    conn_tuple_t t = {};
+    if (!read_conn_tuple(&t, status, sk, pid_tgid, CONN_TYPE_TCP)) {
+        return 0;
+    }
+
+    handle_tcp_stats(&t, status, sk);
```
Comment (p-lambert) [inline pkg/ebpf/c/tracer-ebpf.c]: *if* this turns out to have a perf penalty we could also sample this function call, and only update the RTT value like 10% of the time. this wouldn't work in case of retransmits though.

File: pkg/ebpf/c/tracer-ebpf.c [hunk 8]
```diff
@@ -391,91 +356,72 @@ static void update_conn_stats(
...
-    u32 retransmits,
-    u64 ts) {
-    conn_tuple_t t = {};
-    tcp_stats_t* val;
+static void update_tcp_stats(conn_tuple_t* t, tcp_stats_t stats) {
+    // query stats without the PID from the tuple
+    u32 pid = t->pid;
+    t->pid = 0;
+
+    // initialize-if-no-exist the connetion state, and load it
```
Comment (kevinconaway) [inline pkg/ebpf/c/tracer-ebpf.c]: Do you know why we're doing this?

File: pkg/ebpf/c/tracer-ebpf.h [hunk 9]
```diff
@@ -9,7 +9,8 @@ static const __u8 GUESS_FAMILY = 2;
 static const __u8 GUESS_SPORT = 3;
 static const __u8 GUESS_DPORT = 4;
 static const __u8 GUESS_NETNS = 5;
-static const __u8 GUESS_DADDR_IPV6 = 6;
```
[the remark this answers is not in the document]
Comment (p-lambert) [reply pkg/ebpf/c/tracer-ebpf.h]: No, this is only used internally in the offset guessing process.

File: pkg/ebpf/c/tracer-ebpf.c (hunk 8 above)
Comment (p-lambert) [reply pkg/ebpf/c/tracer-ebpf.c]: I don't have any context, but looking at the code I'd say this is safer from a concurrency standpoint. Otherwise you might have two threads creating a new `tcp_stats_t` object concurrently and not properly aggregating the retransmits.

```
tcp_stats_t* val = bpf_map_lookup_elem(&tcp_stats, t);
t->pid = pid;
    if (val == NULL) {
        return;
    }

File: pkg/ebpf/c/tracer-ebpf.c (hunk 7 above)
Comment (sunhay) [reply pkg/ebpf/c/tracer-ebpf.c]: `bpf_get_prandom_u32()` can help us here

File: pkg/ebpf/event_common.go [hunk 10]
```diff
@@ -108,8 +112,17 @@ func (c ConnectionStats) String() string {
...
+			", %d retransmits (+%d), RTT %s",
+			c.MonotonicRetransmits, c.LastRetransmits,
+			time.Duration(c.RTT)*time.Microsecond,
```
Comment (sunhay) [inline pkg/ebpf/event_common.go]: Should we include variance as well?

Comment (sunhay) [reply pkg/ebpf/event_common.go]: `RTT %s (± %s)`

File: pkg/ebpf/event_common.go [hunk 11]
```diff
@@ -97,8 +101,8 @@ type ConnectionStats struct {
 }
 
 func (c ConnectionStats) String() string {
-	return fmt.Sprintf(
-		"[%s] [PID: %d] [%v:%d ⇄ %v:%d] (%s) %d bytes sent (+%d), %d bytes received (+%d), %d retransmits (+%d)",
+	str := fmt.Sprintf(
+		"[%s] [PID: %d] [%v:%d ⇄ %v:%d] (%s) %d bytes sent (+%d), %d bytes received (+%d)",
```
Comment (sunhay) [inline pkg/ebpf/event_common.go]: I think it's worth importing a bytes to human readable converter to save space and simplify `%d bytes sent` and `%d bytes received`

Something like https://github.com/cloudfoundry/bytefmt/blob/master/bytes.go 

We're probably already importing and using something similar in this repo

File: pkg/ebpf/offsetguess.go [hunk 12]
```diff
@@ -218,6 +289,28 @@ func checkAndUpdateCurrentOffset(module *elf.Module, mp *elf.Map, status *tracer
...
+		// We know that these two fields are always next to each other, 4 bytes apart:
+		// https://elixir.bootlin.com/linux/v4.6/source/include/linux/tcp.h#L232
+		// rtt -> srtt_us
+		// rtt_var -> mdev_us
+		status.offset_rtt++
+		status.offset_rtt_var = status.offset_rtt + 4
+
+		// For now we'll tolerate the case where we can't find the offsets for RTT metrics
+		if status.offset_rtt > thresholdInetSock {
+			log.Warn("could not guess offsets for TCP RTT fields. moving on.")
```
Comment (sunhay) [inline pkg/ebpf/offsetguess.go]: 🙇

File: pkg/ebpf/tracer.go [hunk 13]
```diff
@@ -418,17 +418,17 @@ func (t *Tracer) getTCPStats(mp *bpflib.Map, tuple *ConnTuple, seen map[ConnTupl
 	pid := tuple.pid
 	tuple.pid = 0
 
-	// This is required to avoid (over)reporting stats for connections sharing the same socket.
+	_ = t.m.LookupElement(mp, unsafe.Pointer(tuple), unsafe.Pointer(stats))
```
Comment (sunhay) [inline pkg/ebpf/tracer.go]: Why was this moved up here? non-noticeable performance gain to short circuit if it was already seen?

**Review by sunhay — DISMISSED** [review]
Few small comments - looks good.

File: pkg/ebpf/event_common.go (hunk 11 above)
Comment (p-lambert) [reply pkg/ebpf/event_common.go]: 👍

File: pkg/ebpf/tracer.go (hunk 13 above)
Comment (p-lambert) [reply pkg/ebpf/tracer.go]: basically you can't do an early return anymore in the context of PID collision because we still want to fetch the `TCPStats` and return the RTT metrics.

File: pkg/ebpf/event_common.go (hunk 11 above)
Comment (p-lambert) [reply pkg/ebpf/event_common.go]: indeed we're importing `github.com/dustin/go-humanize`. I'll update it the log entry to use it.

**Comment by sunhay** [comment]
We should add a valid changelog to this using `reno`

File: pkg/ebpf/event_common.go (hunk 10 above)
Comment (p-lambert) [reply pkg/ebpf/event_common.go]: TBH I'm not sure if this is a super rigorous way to represent it, but I think it's a nice to have. I noticed that `ss -i -t` tool prints out `rtt/rtt_var` but I find it even more confusing.

**Review by sunhay — DISMISSED** [review]

**Review by sunhay — APPROVED** [review]
